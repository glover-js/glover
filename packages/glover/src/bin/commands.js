// @flow
import type { Command, Yargs } from 'types/Yargs';

export const brew: Command = {
  cmd: ['brew [recipe]', '$0'],
  desc: 'Brew a recipe',
  builder: (yargs: Yargs) => {
    yargs.positional('recipe', {
      type: 'string',
      describe: 'The recipe you want to brew',
    });
    yargs.option('skipInstall', {
      type: 'boolean',
      describe: 'Skip npm/yarn install',
      default: false,
    });
    yargs.option('skipIngredient', {
      type: 'array',
      describe: 'Skip ingredients in the recipe',
      default: [],
    });
    yargs.option('skipStep', {
      type: 'array',
      describe: 'Skip steps in the recipe',
      default: [],
    });
  },
};

export const pick: Command = {
  cmd: 'pick [ingredients...]',
  desc: 'Pick ingredients from a recipe',
  builder: (yargs: Yargs) => {
    yargs.positional('ingredients', {
      type: 'array',
      describe: 'The ingredients you want to pick',
    });
    yargs.option('prompt', {
      alias: 'p',
      type: 'boolean',
      describe: '',
      default: false,
    });
    yargs.option('output', {
      alias: 'o',
      type: 'string',
      describe: 'Output path',
      default: 'cwd',
    });
  },
};

export const compose: Command = {
  cmd: 'compose [name] [ingredients...]',
  desc: 'Compose a new recipe',
  builder: (yargs: Yargs) => {
    yargs.positional('name', {
      type: 'string',
      describe: 'The name of your recipe',
    });
    yargs.positional('ingredients', {
      type: 'array',
      describe: 'Ingredients you want to include in your recipe',
    });
    yargs.option('extend', {
      alias: 'e',
      type: 'array',
      describe: 'Recipe(s) you want to extend',
      default: [],
    });
    yargs.option('output', {
      alias: 'o',
      type: 'string',
      describe: 'Output path',
      default: 'cwd',
    });
  },
};

export function registerCommands(yargs: Yargs): void {
  const commands = [brew, pick, compose];
  commands.map((cmd: Command) => yargs.command(cmd.cmd, cmd.desc, cmd.builder));
}
