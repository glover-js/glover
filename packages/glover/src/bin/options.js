// @flow
import type { Option } from 'types/Yargs';

export const dryRun: Option = {
  key: 'dryRun',
  options: {
    type: 'boolean',
    describe: 'Run the command without generating any files',
    default: false,
  },
};

export const silent: Option = {
  key: 'silent',
  options: {
    type: 'boolean',
    describe: 'Prevent the command from printing anything to the console',
    default: false,
  },
};

export const noPrompt: Option = {
  key: 'noPrompt',
  options: {
    type: 'boolean',
    describe: 'Prevent the command from prompting for values',
    default: false,
  },
};

export const debug: Option = {
  key: 'debug',
  options: {
    type: 'boolean',
    describe: 'Print debugging info',
    default: false,
  },
};
