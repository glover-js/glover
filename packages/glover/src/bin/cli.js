// @flow
import type { Argv } from 'types/Yargs';
import type { ResolveFn, RejectFn } from 'types/Promise';
import type { RunConfig } from 'types/config';

import yargs from 'yargs';
import { registerCommands } from './commands';
import { registerOptions } from './options';
import glover from '..';

export function parseArgv(rawArgv: Array<string>): Promise<RunConfig> {
  return new Promise((resolve: ResolveFn<RunConfig>, reject: RejectFn) => {
    registerCommands(yargs);
    registerOptions(yargs);
    yargs.scriptName('glover');
    yargs.wrap(120);
    yargs.parse(rawArgv.slice(2), (err: Error, argv: Argv, output: string) => {
      if (err) {
        return reject(err);
      }
      return resolve({
        argv,
        // $FlowFixMe
        command: argv._[0] || 'brew',
        output,
        rawArgv,
      });
    });
  });
}

if (process.env.NODE_ENV !== 'test') {
  parseArgv(process.argv)
    .then((runConfig: RunConfig) => {
      if (runConfig.output) {
        console.log(runConfig.output); // eslint-disable-line no-console
        process.exit(0);
      }
      return glover.run(runConfig);
    })
    .catch((err: Error) => {
      console.log(err); // eslint-disable-line no-console
      process.exit(1);
    });
}
