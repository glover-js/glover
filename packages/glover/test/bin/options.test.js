import * as options from '../../src/bin/options';

describe('Tests for options', () => {
  it('exposes the correct options', () => {
    expect(options).toMatchSnapshot();
  });
});
