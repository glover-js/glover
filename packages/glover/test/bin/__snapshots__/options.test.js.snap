// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for options exposes the correct options 1`] = `
Object {
  "debug": Object {
    "key": "debug",
    "options": Object {
      "default": false,
      "describe": "Print debugging info",
      "type": "boolean",
    },
  },
  "dryRun": Object {
    "key": "dryRun",
    "options": Object {
      "default": false,
      "describe": "Run the command without generating any files",
      "type": "boolean",
    },
  },
  "noPrompt": Object {
    "key": "noPrompt",
    "options": Object {
      "default": false,
      "describe": "Prevent the command from prompting for values",
      "type": "boolean",
    },
  },
  "registerOptions": [Function],
  "silent": Object {
    "key": "silent",
    "options": Object {
      "default": false,
      "describe": "Prevent the command from printing anything to the console",
      "type": "boolean",
    },
  },
}
`;
